nova/scheduler/filters/shard_filter.py

```diff
@@ -113,6 +113,28 @@ def host_passes(self, host_state, spec_obj):
         if nova_utils.is_baremetal_flavor(spec_obj.flavor):
             return True
 
+        host_shard_aggrs = [aggr for aggr in host_state.aggregates
+                            if aggr.name.startswith(self._SHARD_PREFIX)]
+
+        host_shard_names = set(aggr.name for aggr in host_shard_aggrs)
+        if not host_shard_names:
+            LOG.error('%(host_state)s is not in an aggregate starting with '
+                      '%(shard_prefix)s.',
+                      {'host_state': host_state,
+                       'shard_prefix': self._SHARD_PREFIX})
+            return False
+
+        # forbid changing the shard of an instance
+        instance_host = spec_obj.get_scheduler_hint('source_host')
+        host_shard_hosts = set()
+        host_shard_hosts.update(*(aggr.hosts for aggr in host_shard_aggrs))
```

Comment:
Why not `set(aggr.hosts for aggr in host_shard_aggrs)`? I mean, it works the other way too, but...

Reply:
It's not the same:
```
>>> a = [['host1'], ['host2', 'host1']]
>>> set(b for b in a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'list'
```

and if you mean `set(*(aggr.hosts for aggr in host_shard_aggrs))`
```
>>> set(*(b for b in a))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: set expected at most 1 arguments, got 2
```

What would work:
```
>>> set(itertools.chain(*(b for b in a)))
set(['host2', 'host1'])
```

But that's long and even worse to read.

Reply:
ah I missed the unpacking. nevermind then. :)